### PR TITLE
Users api

### PR DESCRIPTION
**Pull request by nikhitashanker**
Added initial route implementations for users API

### CODE REVIEW

File: server/routes/users.js [hunk 1]
```diff
@@ -2,14 +2,272 @@
...
+    }
+    res.send(result.rows);
+  });
+});
+
+
+//having an issue with concat
+/*
+//GET all user names.
+router.get('/', function(req, res, next) {
```
Comment (TrinaKat) [inline server/routes/users.js]: I recall resolving this problem in person, @nikhitashanker can you push the fix?

File: server/routes/users.js [hunk 2]
```diff
@@ -2,14 +2,272 @@
...
+
 // Returns test string to verify that Users server is running.
 router.get('/ping', function(req, res, next) {
   res.send('pong - Users API');
 });
 
 // GET users listing.
 router.get('/', function(req, res, next) {
-    // TODO
+    db.query('SELECT id FROM swe_user', [], (err, result) => {
```
Comment (TrinaKat) [inline server/routes/users.js]: In practice we probably want SELECT *

File: server/routes/users.js [hunk 3]
```diff
@@ -2,14 +2,272 @@
...
+    }
+    res.send(result.rows);
+  });
+});
+*/
+
+
+//GET all user emails.
+router.get('/emails', function(req, res, next) {
+
```
Comment (TrinaKat) [inline server/routes/users.js]: Repeated comment: remove unnecessary whitespace, maintain consistent indentations

File: server/routes/users.js [hunk 4]
```diff
@@ -2,14 +2,272 @@
...
+    if (err) {
+        return next(err);
+    }
+    res.send(result.rows);
+  });
+});
+
+
+//Get user info by user_id
+router.get('/<user_id>', function(req, res, next) {
```
Comment (TrinaKat) [inline server/routes/users.js]: @nikhitashanker push fix discussed in person (change route to /:user_id/id)

File: server/routes/users.js [hunk 5]
```diff
@@ -2,14 +2,272 @@
...
+  const id = req.params.user_id;
```
Comment (TrinaKat) [inline server/routes/users.js]: Nit: keep variable names consistent (e.g. if parameter is `user_id`, use variable `user_id`)

File: server/routes/users.js [hunk 6]
```diff
@@ -2,14 +2,272 @@
...
+        return next(err);
+    }
+    res.send(result.rows);
+  });
+});
+
+//Get a user's past events-HELP NEEDED
+router.get('/:user_id/past', function(req, res, next) {
+  const id = req.params.user_id;
+  db.query('SELECT event_id FROM event_checkin INNER JOIN event ON event_checkin.event_id = event.fb_id WHERE event_checkin.user_id = 3 AND event.ends_at < now() UNION SELECT event_id FROM event_registration INNER JOIN event ON event_registration.event_id = event.fb_id WHERE event_registration.user_id = 3 AND event.ends_at < now()', [], (err, result) => {
```
Comment (TrinaKat) [inline server/routes/users.js]: Reformat this to be readable, test the route to make sure query works as expected, and use the parameter value.

File: server/routes/users.js [hunk 7]
```diff
@@ -2,14 +2,272 @@
...
+        return next(err);
+    }
+    res.send(result.rows);
+  });
+});
+
+//Get a user's favorite events
+//HELP NEEDED
+/*
+router.get('/search', function(req, res, next) {
```
Comment (TrinaKat) [inline server/routes/users.js]: @nikhitashanker use fix discussed in person. 

Note: I see you have this commented code and the route below as well. They have the same base route and should be the same function with checks to see whether the query parameter is set/valid or not.

File: server/routes/users.js [hunk 8]
```diff
@@ -2,14 +2,272 @@
...
+ if (admin){
+  	 db.query('SELECT id FROM swe_user WHERE is_admin = $1', [admin], (err, result) => {
+    	if (err) {
+        	return next(err);
+    	}
+    res.send(result.rows);
+  });
+ }
+});
+
```
Comment (TrinaKat) [inline server/routes/users.js]: Nit: Remove miscellaneous whitespace

File: server/routes/users.js [hunk 9]
```diff
@@ -2,14 +2,272 @@
...
+  //Get all users interested in a given company.
+  if (cid){
+  	db.query('SELECT user_id FROM user_company_rank WHERE company_id = $1', [cid], (err, result) => {
+    	if (err) {
+        	return next(err);
+    	}
+    res.send(result.rows);
+  });
+ }
+ //Get all users of the given occupation
```
Comment (TrinaKat) [inline server/routes/users.js]: Nit: code blocks should be separated by line, comments preceded by line

File: server/routes/users.js [hunk 10]
```diff
@@ -2,14 +2,272 @@
...
+  db.query('SELECT id FROM swe_user WHERE email ILIKE '%$1%'', [email], (err, result) => {
+    if (err) {
+        return next(err);
+    }
+    res.send(result.rows);
+  });
 });
 
+
+router.get('/filter', function(req, res, next) {
```
Comment (TrinaKat) [inline server/routes/users.js]: This route is probably one of the more difficult ones. Glad to see you combined all the options into one route, but the logic should be changed. Luckily, with the users API the filters are all independent (afaik), so the filters should have an AND relationship with each other when fetching results from the database. Think on this and how best to do this without too much redundancy, but don't overthink it. It just involves modifying <10 lines

File: server/routes/users.js [hunk 11]
```diff
@@ -2,14 +2,272 @@
 var express = require('express');
```
Comment (TrinaKat) [inline server/routes/users.js]: Just as a general note, it seems like some of the indents are made of spaces and some are tabs. That doesn't affect much for JS implementation, but for languages like Python which are whitespace-dependent there can be issues. Also, developers will have different tab-to-space conversions (e.g. on my computer a tab can be the width of 2 spaces, while on Swetha's it may be 4 spaces).

TLDR: a non-issue for now, but something to keep in mind. I'm going to suggest using the 2 space width though since the default files use that.

File: server/routes/users.js [hunk 12]
```diff
@@ -2,14 +2,272 @@
...
 // Returns test string to verify that Users server is running.
 router.get('/ping', function(req, res, next) {
   res.send('pong - Users API');
 });
 
 // GET users listing.
 router.get('/', function(req, res, next) {
-    // TODO
+    db.query('SELECT id FROM swe_user', [], (err, result) => {
+    if (err) {
```
Comment (TrinaKat) [inline server/routes/users.js]: Honestly these 4 lines are so repetitive and annoying. I'd just compress it into the following to cut down on wasted space:

```
if (err) return next(err);
res.send(result.rows);
```

File: server/routes/users.js [hunk 13]
```diff
@@ -2,14 +2,272 @@
...
+        return next(err);
+    }
+    res.send(result.rows);
+  });
+});
+*/
+
+
+//GET all user emails.
```
Comment (TrinaKat) [inline server/routes/users.js]: Styling comment: there should be a space between '//' and the comment text

I like how you used CAPS with the http method, try to do this consistently in your comments. (e.g. do GET instead of Get)

File: server/routes/users.js [hunk 14]
```diff
@@ -2,14 +2,272 @@
...
+    }
+    res.send(result.rows);
+  });
+});
+*/
+
+//Get all users whose emails contain the substring email
+//HELP NEEDED
+router.get('/search', function(req, res, next) {
+  const email = req.query.email;
```
Comment (TrinaKat) [inline server/routes/users.js]: Any query parameters must be validated before use

File: server/routes/users.js (hunk 2 above)
Comment (kayliebair) [reply server/routes/users.js]: has been changed to SELECT*

Comment (kayliebair) [reply server/routes/users.js]: has been changed to SELECT *

File: server/routes/users.js [hunk 15]
```diff
@@ -18,7 +18,7 @@ router.get('/', function(req, res, next) {
         return next(err);
     }
     res.send(result.rows);
-  });
```
Comment (TrinaKat) [inline server/routes/users.js]: Seems like there was some confusion, but this should be indented to match the corresponding opening brackets.

e.g.
```
{
    {
         {
         }
    }
    {
    }
}
```
